File: python/navigation/local_planning/band/_band.py
"""Arc-length polyline geometry for the band family (Elastic Bands, TEB).

Both members represent their state as a polyline (bubble centers / poses) that
gets deformed against the obstacle field each tick, so both need arc-length
sampling/resampling of a polyline (band initialization / TEB's progress
projection). Kept as free functions, mirroring the `local_planning/_geometry.py`
precedent for machinery shared across a family's planners rather than owned by
one. (Nearest-occupied lookup, once band-only, now lives in the category-shared
`_geometry.py` since the predictive family reuses it too.)
"""

from __future__ import annotations

import math
from collections.abc import Sequence

from navigation.core.types import Point

from .._geometry import sq_dist


def point_at_arclength(points: Sequence[Point], s: float) -> Point:
    """Point on polyline ``points`` at arc-length ``s`` from its start;
    clamped to the last point once ``s`` reaches or exceeds the total length."""
    if len(points) == 1 or s <= 0.0:
        return points[0]
    remaining = s
    for i in range(len(points) - 1):
        a, b = points[i], points[i + 1]
        seg_len = math.sqrt(sq_dist(a, b))
        if remaining <= seg_len:
            if seg_len < 1e-12:
                return a
            t = remaining / seg_len
            return (a[0] + t * (b[0] - a[0]), a[1] + t * (b[1] - a[1]))
        remaining -= seg_len
    return points[-1]


def resample_polyline(points: Sequence[Point], spacing: float) -> list[Point]:
    """Re-samples ``points`` at even ``spacing`` arc-length intervals, exactly
    preserving the start and end points. A degenerate (near-zero-length) input
    collapses to just those two endpoints.

    Segment count follows the ``max(1, round(total / spacing))`` convention
    already used for distance-to-step-count conversion elsewhere in the
    codebase (`global_planning/sampling/eit_star.py`), so this stays on the
    same rounding contract Python/C++ parity already relies on.
    """
    if len(points) < 2:
        return list(points)
    total = 0.0
    for i in range(len(points) - 1):
        total += math.sqrt(sq_dist(points[i], points[i + 1]))
    if total < 1e-12:
        return [points[0], points[-1]]
    n_segments = max(1, round(total / spacing))
    step = total / n_segments
    out = [points[0]]
    for k in range(1, n_segments):
        out.append(point_at_arclength(points, k * step))
    out.append(points[-1])
    return out

File: cpp/src/local_planning/band/band.cpp
#include "navigation/local_planning/band/band.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

#include "navigation/local_planning/geometry.hpp"

namespace navigation::local_planning {

std::pair<std::optional<core::Point>, double> nearest_occupied(const core::ObstacleQuery& space,
                                                                const core::Point& p,
                                                                double radius) {
  std::optional<core::Point> best;
  double best_sq = std::numeric_limits<double>::infinity();
  for (const core::Point& o : space.occupied_within(p, radius)) {
    double d = sq_dist(p, o);
    if (d < best_sq) {
      best_sq = d;
      best = o;
    }
  }
  if (!best) return {std::nullopt, std::numeric_limits<double>::infinity()};
  return {best, std::sqrt(best_sq)};
}

core::Point point_at_arclength(const std::vector<core::Point>& points, double s) {
  if (points.size() == 1 || s <= 0.0) return points[0];
  double remaining = s;
  for (size_t i = 0; i + 1 < points.size(); ++i) {
    const core::Point& a = points[i];
    const core::Point& b = points[i + 1];
    double seg_len = std::sqrt(sq_dist(a, b));
    if (remaining <= seg_len) {
      if (seg_len < 1e-12) return a;
      double t = remaining / seg_len;
      return core::Point{a.x + t * (b.x - a.x), a.y + t * (b.y - a.y)};
    }
    remaining -= seg_len;
  }
  return points.back();
}

std::vector<core::Point> resample_polyline(const std::vector<core::Point>& points,
                                           double spacing) {
  if (points.size() < 2) return points;
  double total = 0.0;
  for (size_t i = 0; i + 1 < points.size(); ++i) {
    total += std::sqrt(sq_dist(points[i], points[i + 1]));
  }
  if (total < 1e-12) return {points.front(), points.back()};
  // Same max(1, round(total/spacing)) contract as _band.py. Python's round() is
  // half-to-even, and std::lround is half-away-from-zero -- they disagree exactly at
  // .5 ties, so use nearbyint (FE_TONEAREST default = ties-to-even) to keep the
  // segment count bit-identical with Python even on that measure-zero boundary.
  long n_segments = std::max(1L, static_cast<long>(std::nearbyint(total / spacing)));
  double step = total / static_cast<double>(n_segments);
  std::vector<core::Point> out;
  out.reserve(static_cast<size_t>(n_segments) + 1);
  out.push_back(points.front());
  for (long k = 1; k < n_segments; ++k) {
    out.push_back(point_at_arclength(points, static_cast<double>(k) * step));
  }
  out.push_back(points.back());
  return out;
}

}  // namespace navigation::local_planning
